### ISSUE TITLE
Why "valid_acc" always better than "train_acc"?

### ISSUE THREAD

**Issue by Fanxingye** [awslabs/autogluon#1147]
In the summary of different task logs,  "valid_acc" always better than "train_acc", it seems unreasonable.
These are logs of different tasks:
![image](https://user-images.githubusercontent.com/50837470/119929306-d5146e00-bfaf-11eb-86d5-9f2bf13881dc.png)
![image](https://user-images.githubusercontent.com/50837470/119929360-f1180f80-bfaf-11eb-89ae-acd166a8f356.png)

**Comment by talhaanwarch** [comment]
I did not observe this behavior in Vision task
![image](https://user-images.githubusercontent.com/37379131/120020007-98259700-c002-11eb-8268-69c455905e62.png)

**Comment by Fanxingye** [comment]
> I did not observe this behavior in Vision task
> ![image](https://user-images.githubusercontent.com/37379131/120020007-98259700-c002-11eb-8268-69c455905e62.png)

But it always appears in all my Vision tasks, which dataset did you use?

**Comment by talhaanwarch** [comment]
Are you defining tuning set explicitly in predictor function or AG split it self

**Comment by jianzhnie** [comment]
Yes, I did not defin my own val_data, just use the tuning data in predictor function split it self

**Comment by jianzhnie** [comment]
Yes, I did not defin my own val_data, just use the tuning data in predictor function split it self

**Comment by Fanxingye** [comment]
> Are you defining tuning set explicitly in predictor function or AG split it self

Yes, I did not define my own val_data, just use the tuning data in predictor function split it self, Is there any different between these two mode？

**Comment by kevin666aa** [comment]
![image](https://user-images.githubusercontent.com/32823396/120131774-a559a600-c1fb-11eb-850a-b06ff89f5400.png)
I used a dataset from kaggle, [the yoga posts]( https://www.kaggle.com/shrutisaxena/yoga-pose-image-classification-dataset?select=dataset). And I have observed the same problem. 

As you can see, the training accuracy is more than 20% less than validation accuracy, and when I used a self-split test set, I got accuracy similar to the training.

Would you please look into the problem? Maybe test with this kaggle dataset?
PS: My autogluon version is 0.2.0.

**Comment by Innixma** [comment]
Thanks for reporting! This is definitely strange and might have to do with some kind of index misalignment in how training data is scored, hard to tell without deep diving. We will plan to look into this issue and resolve it for v0.3.

**Comment by Innixma** [comment]
@Fanxingye / @kevin666aa Could you provide a kaggle notebook / colab notebook that reproduces the issue reliably? This would help us debug.

**Comment by jianzhnie** [comment]
> @Fanxingye / @kevin666aa Could you provide a kaggle notebook / colab notebook that reproduces the issue reliably? This would help us debug.

Here is a demo from the docs of autogluon, I tested it  and observed the same problem. https://auto.gluon.ai/stable/tutorials/image_prediction/hpo.html

**Comment by Innixma** [comment]
Thanks, it does appear to be quite widespread. The primary maintainer of the vision portion of the codebase is currently on vacation this week, but we will try to resolve as much as we can once he is back.

**Comment by talhaanwarch** [comment]
> ![image](https://user-images.githubusercontent.com/32823396/120131774-a559a600-c1fb-11eb-850a-b06ff89f5400.png)
> I used a dataset from kaggle, [the yoga posts](https://www.kaggle.com/shrutisaxena/yoga-pose-image-classification-dataset?select=dataset). And I have observed the same problem.
> 
> As you can see, the training accuracy is more than 20% less than validation accuracy, and when I used a self-split test set, I got accuracy similar to the training.
> 
> Would you please look into the problem? Maybe test with this kaggle dataset?
> PS: My autogluon version is 0.2.0.

Hi, @zhreshold and @Innixma  I would like to hear your opinion on @kevin666aa comment as he is saying he got similar accuracy when he do splitting him and much lesser accuracy when splitting is performed by predictor. Or may I misunderstood what he is saying

**Comment by talhaanwarch** [comment]
> > ![image](https://user-images.githubusercontent.com/32823396/120131774-a559a600-c1fb-11eb-850a-b06ff89f5400.png)
> > I used a dataset from kaggle, [the yoga posts](https://www.kaggle.com/shrutisaxena/yoga-pose-image-classification-dataset?select=dataset). And I have observed the same problem.
> > As you can see, the training accuracy is more than 20% less than validation accuracy, and when I used a self-split test set, I got accuracy similar to the training.
> > Would you please look into the problem? Maybe test with this kaggle dataset?
> > PS: My autogluon version is 0.2.0.
> 
> Hi, @zhreshold and @Innixma I would like to hear your opinion on @kevin666aa comment as he is saying he got similar accuracy when he do splitting himself and much lesser accuracy when splitting is performed by predictor. Or may I misunderstood what he is saying

**Comment by Innixma** [comment]
@zhreshold It may be worth trying to replicate the issue shown by @kevin666aa, this feels like an actual bug to me.

**Comment by talhaanwarch** [comment]
I am not sure, but I would suggest to double check if data is shuffled before splitting

**Comment by talhaanwarch** [comment]
colab notebook
https://colab.research.google.com/drive/1_tZ9y-gmzelaLLLX6eSO_NLUg-ksD6qy?usp=sharing

**Comment by Innixma** [comment]
Deep dived into this and it is indeed a bug.

When validation data isn't explicitly specified to `predictor.fit()`, 100% of data is used for both training and validation.

The reason why is not so simple however... Everything looks like its going correctly until deep in GluonCV code when the ImageDatasets are transformed into their mxnet versions via `train_data.to_mxnet()` and `val_data.to_mxnet()`. The output of `to_mxnet()` somehow returns the full original data prior to splitting for both.

Here is a colab notebook showing the bug:

https://colab.research.google.com/drive/1pJScwfjj1lNIC07AR4jJxLw__9of639K?usp=sharing

**Comment by Innixma** [comment]
Opened #1168 which fixes ImagePredictor functionality.
